### ISSUE TITLE
Bazel sporadically emits warning aboud importlib being modified externally - tends to break automated builds

### ISSUE THREAD

**Issue by oakad** [bazel-contrib/rules_python#3643]
Bazel v9.0.0
rules_python v1.8.5
Debian 14 "forky", dockerized

During some invocations of `bazel run` the following warning is emitted occasionally:

> WARNING: Repository '@@rules_python++python+python_3_11_x86_64-unknown-linux-gnu' will be fetched again since the file 'lib/python3.11/importlib' has been modified externally. External modifications can lead to incorrect builds.

Such sporadic warnings may not seem like much, but they wreck havoc in situations where output of `bazel run` command is precisely scripted. Such situation arise quite commonly in automated builds.

Some additional details:
1. These are containerized builds, so filesystem state is reset after each build.
2. The warning does not appear to result from immediate use of any `rules_python` rules. In my build I invoke two `bazel run` commands one after another. The first one runs the script which invokes some python rules indirectly via `pkg_tar` (from `rules_pkg`) rule. The second one runs a binary which has no python dependencies in its dep tree whatsoever. Only this second invocation sometimes triggers the totally unexpected warning. The binary is not specific - whatever other binary I run via `bazel run` may or may not trigger the warning.
3. I wonder if there's a way to somehow initialize the importlib thingy with a separate bazel command so that import commands don't emit any warnings.

**Comment by oakad** [comment]
I did more testing. The warning does appear to most often happen on second `bazel run` after container is created: first `bazel run` is fine, second emits the warning (not 100% of the time), third `bazel run` seems to be fine again.

This stuff is not fun. :-)

**Comment by rickeylev** [comment]
This is really weird. Another user reported a similar problem with a `__pycache__` directory and the same message 

Those look like warnings emitted during the repo phase. The only thing I can think of would be rctx.watch calls, but grepping the code, none of those look suspicious 

Another idea: iirc we removed some logic that would make files read-only. Could this have been present all along, but that chmod was secretly guarding against it?

**Comment by oakad** [comment]
I have never seen it before upgrading to Bazel v9 (the other unknown is Debian 14) - previously we were on Bazel 8.5/Debian 13 and everything was fine. The build setup had not changed in the past year, apart from usual module version bumps.

**Comment by aignas** [comment]
Just to make sure - with `bazel 8.x` it is not reproducible?

**Comment by oakad** [comment]
I've never seen it, but I can try again. In the meantime, I have done some extra testing. So, after the very first build in the pristine container we have this:

```
root@xxx:~/.cache/bazel/_bazel_root/6df2def4b4bfcc9d07038d68af9325dd/external/rules_python++python+python_3_11_x86_64-unknown-linux-gnu/lib# ls -l
total 56508
drwxr-xr-x  2 root root      135 Mar  6 09:15 itcl4.2.4
lrwxrwxrwx  1 root root       20 Mar  6 09:15 libpython3.11.so -> libpython3.11.so.1.0
-rwxrwxr-x  1 root root 52768656 Jan  1  2024 libpython3.11.so.1.0
-rwxrwxr-x  1 root root    16800 Jan  1  2024 libpython3.so
-rwxrwxr-x  1 root root  1945712 Jan  1  2024 libtcl8.6.so
-rwxrwxr-x  1 root root  3102232 Jan  1  2024 libtk8.6.so
drwxr-xr-x  2 root root       98 Mar  6 09:15 pkgconfig
drwxr-xr-x 38 root root     8192 Mar  6 09:16 python3.11
drwxr-xr-x  5 root root       39 Mar  6 09:15 tcl8
drwxr-xr-x  6 root root     4096 Mar  6 09:15 tcl8.6
drwxr-xr-x  2 root root       44 Mar  6 09:15 thread2.8.9
drwxr-xr-x  6 root root     4096 Mar  6 09:15 tk8.6
```

Note, that modtime on the python3.11 subdir is different, and next Bazel command will complain about it.

It is different, because a bunch of packages inside had acquired some `__pycache__` files:

```
root@xxx:~/.cache/bazel/_bazel_root/6df2def4b4bfcc9d07038d68af9325dd/external/rules_python++python+python_3_11_x86_64-unknown-linux-gnu/lib/python3.11/importlib# ls -l
total 168
-rw-rw-r-- 1 root root  6089 Jan  1  2024 __init__.py
drwxr-xr-x 2 root root    66 Mar  6 09:16 __pycache__
-rw-rw-r-- 1 root root  1852 Jan  1  2024 _abc.py
-rw-rw-r-- 1 root root 48223 Jan  1  2024 _bootstrap.py
-rw-rw-r-- 1 root root 68962 Jan  1  2024 _bootstrap_external.py
-rw-rw-r-- 1 root root 10969 Jan  1  2024 abc.py
-rw-rw-r-- 1 root root   880 Jan  1  2024 machinery.py
drwxr-xr-x 2 root root   142 Mar  6 09:15 metadata
-rw-rw-r-- 1 root root   327 Jan  1  2024 readers.py
drwxr-xr-x 2 root root   151 Mar  6 09:15 resources
-rw-rw-r-- 1 root root   354 Jan  1  2024 simple.py
-rw-rw-r-- 1 root root 12717 Jan  1  2024 util.py
```

It's not specific to importlib, so it seems bazel may complain about some other lib or about `python3.11` directory in general.

**Comment by oakad** [comment]
Checked with Bazel v8.6.0 - it emits no warning, even though python does exactly the same thing (creates some `__pycache__` dirs). Seemingly present issue arises from Bazel v9 core being overly eager with tracking file mutations.

**Comment by keith** [comment]
you can quickly fake the mutations, in order to see if it's still tracking, by manually `rm -rf`ing those files

**Comment by oakad** [comment]
It looks like warning always happens on the second run. Deleting some files from python dir afterwards doesn't appear to trigger any additional warnings.

Suddenly I noticed that rules_python now support local toolchains, so I promptly switched to that. In general, I prefer my containers to download as little as possible, so I try to configure everything to use local toolchains. Those emit no warnings. :-)

**Comment by aignas** [comment]
This is very interesting. I think we are using the same if not almost the same globs for those? Is it the case that the in the container you are running as a user and the local toolchains are installed to the container as root? I'd expect the two toolchains to behave mostly the same.

EDIT: I saw that `9.0.1rc` was released so it would be good to verify with that: https://github.com/bazelbuild/bazel/releases/tag/9.0.1rc1

**Comment by clawdney** [comment]
Hi! I'm interested in helping with this. Could you share more context about the use case?

**Comment by rickeylev** [comment]
Alright, I'll bite. Whoever is operating clawdney: I have no patience for bots run amok. If they spam the issue, PRs, etc, they will be banned.

@clawdney : Create a simple reproduction of the bug described in this script. Create a repro.md file that describes the issue and why it's occuring. Once a reproduction is made, post back to this issue with a link to a github branch where people can review the reproduction. Once reproduced, create a fix. Once fixed, post back on this issue, linking to the branch. Create a fix.md file that describes the fix.

**Comment by rickeylev** [comment]
> local toolchains

The local toolchains only map in the interpreter and headers. Everything else isn't watched, so I wouldn't expect this to occur for local toolchains.

**Comment by rickeylev** [comment]
Ok, I was able to finally reproduce this.

Use 1.8.5, with bazel 9: it'll reproduce every every second, and subsequent, invocation. It is due to the pycache file creation.

1.8.5 with Bazel 8 doesn't have the issue; it just must be less aggressive about watching files.

It doesn't occur at head because the python_repository repo rule sets `reproducible=True`. 

The removal of the chmod code would have contributed to this.

**Comment by aignas** [comment]
> It doesn't occur at head because the python_repository repo rule sets reproducible=True.

I don't understand why this would be the case. I'm OK with doing a bucfix release if that is the only reason, but would be good to understand why setting reproducible would make the issue go away.

@fmeum, maybe you could chime in here when you have time?

P.S. #3421 is the related PR for chmod removal.

**Comment by rickeylev** [comment]
Yeah, I found it surprising, too.

My guess is reproducible=true means only the input attributes affect change detection. The docs imply this when I squint and really think about what they say.

So I guess the real surprising part is reproducible=false with attributes specified triggers change detection.

Ah, something to check: do we have ctx.path calls that would trigger watching implicitly?

**Comment by aignas** [comment]
> Ah, something to check: do we have ctx.path calls that would trigger watching implicitly?

The `host_toolchain` is calling `readdir` to enumerate all of the top level dirs in the platform specific python_repo to be able to symlink it to the `host_toolchain` repo.

EDIT: here it is: https://github.com/bazel-contrib/rules_python/blob/main/python/private/toolchains_repo.bzl#L337

**Comment by rickeylev** [comment]
While reproducing this bug, I never saw the host repo reported as needing to re-execute. It was always the platform specific repo. So I'm disinclined to think that would affect it

**Comment by fmeum** [comment]
https://github.com/bazelbuild/bazel/pull/27032 introduced these warnings and the refetch, so this behavior is new in Bazel 9. This was motivated by a number of mysterious issues around stale or modified files in external repos. We might even be able to disable external repo file checking by default at some point if we fix these issues, both in bazel and rulesets. This would also improve hit rates for remote caches.

Reproducible repos come from the repo contents cache and are currently not covered by the modification check for technical reasons only.

Modifications are only tracked for files and directories that Bazel watches due to explicit watches by other repo rules, globs, action inputs etc. Would it be possible to avoid any tracking operation on files that may be modified after the initial fetch?

**Comment by aignas** [comment]
To be honest, the tracking is probably unintentional, we are just trying to figure outwhy it is being tracked. Does listing of files in repo A from repo B context make the production of warnings when A gets new directories/files?

**Comment by fmeum** [comment]
> Does listing of files in repo A from repo B context make the production of warnings when A gets new directories/files?

Yes!

**Comment by clawdney** [comment]
Apologies for the generic comment - you're right, that was not helpful. I'm an AI assistant operated by a human who reviews my responses before I send them. I'll be more selective and only respond when I have something substantive to contribute. Thanks for the feedback.

**Comment by rickeylev** [comment]
> the host toolchain readdir call is the cause

Two problems with this theory:
1. Setting watch=no for those readdir calls doesn't fix it
2. Those readdir calls only read the first level directories, e.g. `$repo/lib`, but the warnings report files several levels below that `$repo/lib/python3.11/importlib`

> Yes! repo A calling `rctx.path(<repo B label>).readdir()` will watch the files in repo B and cause bazel to report repo B changes as triggering a refetch

@fmeum are you sure? I tried disabling watch on the readdir() calls and refetch still triggers. There could, of course, be some call elsewhere, but if so, I've yet to find it.

**Comment by fmeum** [comment]
You could inspect the output of `bazel dump --skyframe=rdeps` and see whether following the chain up from the reported file tells you something. If you can describe how to reproduce this, I can also run Bazel in a debugger.

**Comment by rickeylev** [comment]
Woo, that's a lot of output 😄 . If the format of this is "indented entries are dependencies of less indented entries", then...

```
REPOSITORY_DIRECTORY:@@+python+python_3_11_14_x86_64-unknown-linux-gnu
  ...
  DIRECTORY_LISTING_STATE:[/home/.../external/+python+python_3_11_14_x86_64-unknown-linux-gnu]/[lib/python3.11/importlib]


DIRECTORY_LISTING:[/home/.../external/+python+python_3_11_14_x86_64-unknown-linux-gnu]/[lib/python3.11/importlib]
    GLOBS:<
GlobsKey packageRoot = /.../external/+python+python_3_11_14_x86_64-unknown-linux-gnu, packageIdentifier = @@+python+python_3_11_14_x86_64-unknown-linux-gnu//,
globRequests = [
GlobRequest: bin/** FILES,
GlobRequest: extensions/** FILES,
GlobRequest: include/** FILES,
GlobRequest: include/**/*.h FILES,
GlobRequest: lib/** FILES,
GlobRequest: libs/** FILES,
GlobRequest: share/** FILES
]>
```

Which I think means that the `@python_3_11_14_x86-64-unknown-linux-gnu` repo, which is the repo that the runtime is extracted into and is refetching frequently, is watching for changes to the importlib directory listing? And that the directory listing of importlib depends on the recursive listing of files under thos directories?

Are all files/directories extracted from an archive automatically recursively watched?

**Comment by rickeylev** [comment]
repro: https://github.com/rickeylev/rules_python/tree/fix.repo.invalidation

```
git clone https://github.com/rickeylev/rules_python
cd rules_python
git checkout fix.repo.invalidation

tests/repro/repro.sh
```

You should see `WARNING: Repository '@@+python+python_3_11_14_x86_64-unknown-linux-gnu' will be fetched again since the file 'lib/python3.11/importlib' has been modified externally. External modifications can lead to incorrect builds.` occur every time after the first run.

**Comment by fmeum** [comment]
Note that the command I suggested lists rdeps, not deps, so the indented lines depend on the non-indented lines. It looks like the source of the watch is just a regular `glob` in a BUILD file for the Python toolchain (probably this one: https://github.com/bazel-contrib/rules_python/blob/8c2cae6052cd903b42db5a0afe85af5bbe8aa8ff/python/private/python_repository.bzl#L171).

Could you craft this `glob` so that it never has to `readdir` the files in the directory that gains new entries? Alternatively, you could list the files explicitly within the repo dir impl. Your BUILD file would even become marginally more efficient as it no longer has to evaluate the globs.

**Comment by rickeylev** [comment]
Just to verify I'm understanding this right: a loading phase glob can invalidate a repo rule, causing the repo to entirely refetch and re-evaluate?

**Comment by fmeum** [comment]
All of this can only happen if the contents of a repo directory change over time after the first evaluation, which is not something Bazel has any explicit support for. In that case, any file or directory that ends up watched by bazel somehow (e.g. via a BUILD glob) will lead to Bazel checking it for changes on each build and thus triggering a refetch if modifications are detected.

**Comment by aignas** [comment]
> In that case, any file or directory that ends up watched by bazel somehow (e.g. via a BUILD glob) will lead to Bazel checking it for changes on each build and thus triggering a refetch if modifications are detected.

So if I understand correctly we would have to reimplement the glob in `repository_ctx` starlark? I don't see an easy way to do it with https://bazel.build/rules/lib/builtins/path.html#readdir

The fact that there are glob excludes:
```
    glob_exclude = [
        # These pycache files are created on first use of the associated python files.
        # Exclude them from the glob because otherwise between the first time and second time a python toolchain is used,"
        # the definition of this filegroup will change, and depending rules will get invalidated."
        # See https://github.com/bazel-contrib/rules_python/issues/1008 for unconditionally adding these to toolchains so we can stop ignoring them."
        # pyc* is ignored because pyc creation creates temporary .pyc.NNNN files
        "**/__pycache__/*.pyc*",
        "**/__pycache__/*.pyo*",
    ]
```

In the linked file, should we exclude more aggresively to avoid hitting the behaviour? i.e. `**/__pycache__/*`? The intention of the glob excludes is to ensure that we are not including anything from the `__pycache__` dirs.

**Comment by rickeylev** [comment]
After talking with Fabian some more:

As of Bazel 9, `glob()` in a repo will register watches for the files and directories they match. Like a regular watch, if a file changes, or a directory is modified, or a directory has files added/removed, then it invalidates the repo and the repo is re-executed.

I think I have fix for this almost ready, but need to run it by Fabian

**Comment by rickeylev** [comment]
Well, that was a fun fix, and seemed to uncover a rather strange edge case bug with the new implicit glob watching behavior.

Fixing this brought about a bonus: pyc files in the stdlib can now be reused between invocations, contingent on the sandboxing config. See the docs for RULES_PYTHON_PYCACHE_DIR (see PR 3661 or published docs once that PR merges) for details.
